client-libraries/rust/src/plugin_utils/ingest.rs

```diff
@@ -0,0 +1,1065 @@
+                            // plugins.ingest.collectors.<full bin filename>
+                            ingest
+                                .importers
+                                .get(file_stem.as_str())
+                                // plugins.ingest.collectors.<stem filename>
+                                .or_else(|| ingest.importers.get(file_stem.alias()))
+                        } else {
+                            None
+                        };
+
+                        if let Some(pim) = plugins_ingest_member {
+                            // If we identified a named toml entry, merge it in to the top level as 'metadata'.
+                            raw_toml.metadata = pim.metadata.clone();
+
+                            if raw_toml.ingest.is_none() {
+                                raw_toml.ingest = Some(Default::default());
+                            }
+                            raw_toml.ingest.as_mut().unwrap().timeline_attributes =
+                                pim.timeline_attributes.clone();
+                        }
+                    }
+                }
+            }
+
+            let r: Result<crate::reflector_config::Config, SemanticErrorExplanation> =
+                raw_toml.try_into();
+            cfg = Some(r.map_err(|semantics| ConfigLoadError::DefinitionSemantics {
+                explanation: semantics.0,
+            })?);
+        }
+
+        let cfg = cfg.unwrap_or_default();
+
+        let mut ingest = cfg.ingest.clone().unwrap_or_default();
+        override_ingest_config_from_env(&mut ingest)?;
+
+        let env_config = envy::from_env::<EnvConfig>()?;
+
+        // Load plugin-specific config from the 'metdata' entry
+        let mut plugin_toml = cfg.metadata.clone();
+        merge_plugin_config_from_env::<T>(env_prefix, map_env_val, &mut plugin_toml)?;
+
+        // deserialize from merged toml values to the actual struct
+        let plugin: T = TomlValue::Table(plugin_toml.into_iter().collect()).try_into()?;
+
+        // syntheisze a uuid runid if none was given
+        let run_id = env_config
+            .modality_run_id
+            .unwrap_or_else(|| uuid::Uuid::new_v4().to_string());
+
+        let client_timeout = env_config
+            .modality_client_timeout
+            .map(Duration::from_secs_f32);
+
+        Ok(Config {
+            ingest,
+            plugin,
+            client_timeout,
+            run_id,
+            time_domain: env_config.modality_time_domain,
+        })
+    }
+
+    /// Connect to the configured Modality backend for ingest,
+    /// authenticate, and return a high-level ingest client.
+    pub async fn connect_and_authenticate(&self) -> Result<Client, Box<dyn std::error::Error>> {
+        let protocol_parent_url = if let Some(url) = &self.ingest.protocol_parent_url {
+            url.clone()
+        } else {
+            Url::parse("modality-ingest://127.0.0.1")?
```

Comment:
Shouldn't we be checking for `MODALITY_HOST` first?

Reply:
That happens in `override_ingest_config_from_env`, which happens before this, so `self.ingest.protocol_parent_url` contains the final answer.